receiver/splunkhecreceiver/receiver.go

```diff
@@ -0,0 +1,319 @@
+	jsonContentType           = "application/json"
+	gzipEncoding              = "gzip"
+	httpContentTypeHeader     = "Content-Type"
+	httpContentEncodingHeader = "Content-Encoding"
+)
+
+var (
+	errNilNextConsumer = errors.New("nil nextConsumer")
+	errEmptyEndpoint   = errors.New("empty endpoint")
+
+	okRespBody               = initJSONResponse(responseOK)
+	invalidMethodRespBody    = initJSONResponse(responseInvalidMethod)
+	invalidContentRespBody   = initJSONResponse(responseInvalidContentType)
+	invalidEncodingRespBody  = initJSONResponse(responseInvalidEncoding)
+	errGzipReaderRespBody    = initJSONResponse(responseErrGzipReader)
+	errReadBodyRespBody      = initJSONResponse(responseErrReadBody)
+	errUnmarshalBodyRespBody = initJSONResponse(responseErrUnmarshalBody)
+	errNextConsumerRespBody  = initJSONResponse(responseErrNextConsumer)
+)
+
+// splunkReceiver implements the component.MetricsReceiver for Splunk HEC metric protocol.
+type splunkReceiver struct {
+	sync.Mutex
+	logger       *zap.Logger
+	config       *Config
+	nextConsumer consumer.MetricsConsumer
+	server       *http.Server
+
+	startOnce sync.Once
+	stopOnce  sync.Once
+}
+
+var _ component.MetricsReceiver = (*splunkReceiver)(nil)
+
+// New creates the Splunk HEC receiver with the given configuration.
+func New(
+	logger *zap.Logger,
+	config Config,
+	nextConsumer consumer.MetricsConsumer,
+) (component.MetricsReceiver, error) {
+	if nextConsumer == nil {
+		return nil, errNilNextConsumer
+	}
+
+	if config.Endpoint == "" {
+		return nil, errEmptyEndpoint
+	}
+
+	r := &splunkReceiver{
+		logger:       logger,
+		config:       &config,
+		nextConsumer: nextConsumer,
+		server: &http.Server{
+			Addr: config.Endpoint,
+			// TODO: Evaluate what properties should be configurable, for now
+			//		set some hard-coded values.
+			ReadHeaderTimeout: defaultServerTimeout,
+			WriteTimeout:      defaultServerTimeout,
+		},
+	}
+
+	return r, nil
+}
+
+// StartMetricsReception tells the receiver to start its processing.
+// By convention the consumer of the received data is set when the receiver
+// instance is created.
+func (r *splunkReceiver) Start(_ context.Context, host component.Host) error {
+	r.Lock()
+	defer r.Unlock()
+
+	err := componenterror.ErrAlreadyStarted
+	r.startOnce.Do(func() {
+		err = nil
+
+		var ln net.Listener
+		// set up the listener
+		ln, err = r.config.HTTPServerSettings.ToListener()
+		if err != nil {
+			err = fmt.Errorf("failed to bind to address %s: %w", r.config.Endpoint, err)
+			return
+		}
+
+		mx := mux.NewRouter()
+		mx.HandleFunc(hecPath, r.handleReq)
+
+		r.server = r.config.HTTPServerSettings.ToServer(mx)
+
+		// TODO: Evaluate what properties should be configurable, for now
+		//		set some hard-coded values.
+		r.server.ReadHeaderTimeout = defaultServerTimeout
+		r.server.WriteTimeout = defaultServerTimeout
+
+		go func() {
+			if errHTTP := r.server.Serve(ln); errHTTP != nil {
+				host.ReportFatalError(errHTTP)
```

Comment:
I seem to recall Serve returning ErrServerClosed on shutdown?

Reply:
straight copy from sfxreceiver, if there's a better way to handle this, let me know.